chemprop/cli/hyperopt.py

```diff
@@ -1,16 +1,348 @@
+    )
+
+    checkpoint_config = CheckpointConfig(
+        num_to_keep=args.num_checkpoints_to_keep,
+        checkpoint_score_attribute="val_loss",
+        checkpoint_score_order=monitor_mode,
+    )
+
+    run_config = RunConfig(
+        checkpoint_config=checkpoint_config,
+    )
+
+    ray_trainer = TorchTrainer(
+        lambda config: train_model(config, args, train_loader, val_loader, logger),
+        scaling_config=scaling_config,
+        run_config=run_config,
+    )
+
+    tune_config = tune.TuneConfig(
+        metric="val_loss",
+        mode=monitor_mode,
+        num_samples=args.num_samples,
+        scheduler=scheduler,
+    )
+
+    tuner = tune.Tuner(
+        ray_trainer,
+        param_space={"train_loop_config": build_search_space(args.search_parameter_keywords, args.epochs)},
+        tune_config=tune_config,
+    )
+
+    return tuner.fit()
+
+def main(args: Namespace):
+
+    format_kwargs = dict(
+        no_header_row=args.no_header_row,
+        smiles_cols=args.smiles_columns,
+        rxn_cols=args.reaction_columns,
+        target_cols=args.target_columns,
+        ignore_cols=args.ignore_columns,
+        weight_col=args.weight_column,
+        bounded=args.loss_function is not None and "bounded" in args.loss_function,
+    )
+    featurization_kwargs = dict(
+        features_generators=args.features_generators, keep_h=args.keep_h, add_h=args.add_h
+    )
+
+    train_data, val_data, test_data = build_splits(args, format_kwargs, featurization_kwargs)
+    train_dset, val_dset, test_dset = build_datasets(args, train_data, val_data, test_data)
+
+    X_d_scaler, V_f_scaler, E_f_scaler, V_d_scaler = normalize_inputs(
+        train_dset, val_dset, args
+    )
+    input_scalers = {"X_d": X_d_scaler, "V_f": V_f_scaler, "E_f": E_f_scaler, "V_d": V_d_scaler}
+
+    if "regression" in args.task_type:
+        scaler = train_dset.normalize_targets()
+        val_dset.normalize_targets(scaler)
+        logger.info(f"Train data: mean = {scaler.mean_} | std = {scaler.scale_}")
+    else:
+        scaler = None
+
+    train_loader = MolGraphDataLoader(
+        train_dset, args.batch_size, args.num_workers, seed=args.data_seed
+    )
+    val_loader = MolGraphDataLoader(val_dset, args.batch_size, args.num_workers, shuffle=False)
+    if test_dset is not None:
+        test_loader = MolGraphDataLoader(
+            test_dset, args.batch_size, args.num_workers, shuffle=False
+        )
+    else:
+        test_loader = None
+
+    torch.manual_seed(args.pytorch_seed)
+
+    model = build_model(args, train_loader.dataset)
+    monitor_mode = "min" if model.metrics[0].minimize else "max"
+
+    results = tune_model(args, train_loader, val_loader, logger, monitor_mode)
+
+    best_result = results.get_best_result()
+    best_config = best_result.config
+    best_checkpoint = best_result.checkpoint  # Get best trial's best checkpoint
```

Comment:
I would be in favor of saving this, as the user could start a training run from this checkpoint, and have the best configuration already set up, right?

Reply:
Thanks! I have added code to save the checkpoint.